node/bft/src/primary.rs

```diff
@@ -955,14 +955,28 @@ impl<N: Network> Primary<N> {
             bail!("Received a batch certificate for myself ({author})");
         }
 
-        // Store the certificate, after ensuring it is valid.
+        // Ensure that the incoming certificate is valid.
+        self.storage.check_incoming_certificate(&certificate)?;
+
+        // Store the certificate, after ensuring it is valid above.
+        // The following call recursively fetches and stores
+        // the previous certificates referenced from this certificate.
+        // It is critical to make the following call this after validating the certificate above.
+        // The reason is that a sequence of malformed certificates,
+        // with references to previous certificates with non-decreasing rounds,
+        // cause the recursive fetching of certificates to crash the validator due to resource exhaustion.
+        // Note that if the following call, if not returning an error, guarantees the backward closure of the DAG
+        // (i.e. that all the referenced previous certificates are in the DAG before storing this one),
+        // then all the validity checks in [`Storage::check_certificate`] should be redundant.
+        // TODO: eliminate those redundant checks
```

Comment:
Are we removing the redundant checks in another PR?

Reply:
Yes, there are currently several redundant checks (not just this one), and they are a bit complicated (I documented them in the code notes doc), so my thinking was to tackle those in a series of PRs each of which makes a very simple change that is easy to evaluate.